### PR TITLE
Selection for wide glyphs

### LINKED ISSUE #935: Fix wide glyph selection
# Summary of the new feature/enhancement

Selection should always expand to cover the entire glyph, regardless of its size. This particularly makes sense for most emoji characters spanning two spaces.

# Proposed technical implementation details (optional)

When selection reaches the left side of a wide glyph, it should expand to the right (unless it's already selected). The same should apply for the reverse.

An important corner case is box selection. Here, the selection should expand for the individual row, NOT the entire selection.

### PR DESCRIPTION
**Pull request by carlos-zamora**
## Summary of the Pull Request

Fixes a bug/feature to select wide glyphs (#935 ). Short summary of it is that selecting emojis and other wide glyphs should encompass the entire glyph, not just half.

Also moved selection logic to new file for cleanliness.

I had to rebase from the ADO branch but it looks good. And tested :)

## PR Checklist
* [x] Closes #935 
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [x] Tests added/passed
* [x] Requires documentation to be updated
* [x] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Detailed Description of the Pull Request / Additional comments
Basically, selection has to be covered in two cases: expand left and expand right. Made 2 similar looking functions.

To the team: this was already somewhat reviewed back before we went on GitHub. So if you have a deja vu feeling, that's why.

### CODE REVIEW

File: src/cascadia/TerminalCore/TerminalSelection.cpp [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    return selectionArea;
+}
+
+// Method Description:
+// - Expands the selection left-wards to cover a wide glyph, if necessary
+// Arguments:
+// - position: the (x,y) coordinate on the visible viewport
+// Return Value:
+// - updated x position to encapsulate the wide glyph
+const SHORT Terminal::_ExpandWideGlyphSelection_Left(const SHORT x_pos, const SHORT y_pos) const noexcept
```
Comment (zadjii-msft) [inline src/cascadia/TerminalCore/TerminalSelection.cpp]: `x_pos` and `y_pos` definitely don't match our naming - `xPos`

File: src/cascadia/TerminalCore/TerminalSelection.cpp [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    return position.X;
+}
+
+// Method Description:
+// - Expands the selection right-wards to cover a wide glyph, if necessary
+// Arguments:
+// - position: the (x,y) coordinate on the visible viewport
+// Return Value:
+// - updated x position to encapsulate the wide glyph
+const SHORT Terminal::_ExpandWideGlyphSelection_Right(const SHORT x_pos, const SHORT y_pos) const noexcept
```
Comment (zadjii-msft) [inline src/cascadia/TerminalCore/TerminalSelection.cpp]: same with these params

**Comment by miniksa** [comment]
Carlos is out of office until Tuesday. It's not fair if this gets marked stale on him given he has full intent on finishing this when he comes back.

**Comment by carlos-zamora** [comment]
Just adding the tests @ msftbot! Calm down!

File: src/cascadia/TerminalCore/TerminalSelection.cpp [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    const COORD endSelectionPositionWithOffset = { _endSelectionPosition.X, temp2 };
+
+    // NOTE: (0,0) is top-left so vertical comparison is inverted
+    const COORD &higherCoord = (selectionAnchorWithOffset.Y <= endSelectionPositionWithOffset.Y) ? selectionAnchorWithOffset : endSelectionPositionWithOffset;
+    const COORD &lowerCoord = (selectionAnchorWithOffset.Y > endSelectionPositionWithOffset.Y) ? selectionAnchorWithOffset : endSelectionPositionWithOffset;
+
+    selectionArea.reserve(lowerCoord.Y - higherCoord.Y + 1);
+    for (auto row = higherCoord.Y; row <= lowerCoord.Y; row++)
+    {
+        SMALL_RECT selectionRow;
```
Comment (dlong11) [inline src/cascadia/TerminalCore/TerminalSelection.cpp]: Not necessary for this PR,  but in cases like this do you guys recommend initializing these structs (when they can't be initialized at declaration time.
Other projects I work on have a rule where this should be SMALL_RECT{};  I think this is a good rule and prevents unintialized vars.
Just checking for future reviews.  @adiviness

Comment (adiviness) [reply src/cascadia/TerminalCore/TerminalSelection.cpp]: I personally prefer initializing any variable with some known value when it's created. I'm sorta hoping there's a static analysis check to enforce this at some point.

File: src/cascadia/TerminalCore/TerminalSelection.cpp [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+{
+    std::vector<SMALL_RECT> selectionArea;
+
+    if (!_selectionActive)
+    {
+        return selectionArea;
+    }
+
+    // Add anchor offset here to update properly on new buffer output
+    SHORT temp1, temp2;
```
Comment (adiviness) [inline src/cascadia/TerminalCore/TerminalSelection.cpp]: [ES.10](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-name-one) recommends one name per declaration.

File: src/cascadia/TerminalCore/TerminalSelection.cpp [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+    THROW_IF_FAILED(ShortAdd(_selectionAnchor.Y, _selectionAnchor_YOffset, &temp1));
+    THROW_IF_FAILED(ShortAdd(_endSelectionPosition.Y, _endSelectionPosition_YOffset, &temp2));
+
+    // create these new anchors for comparison and rendering
+    const COORD selectionAnchorWithOffset = { _selectionAnchor.X, temp1 };
+    const COORD endSelectionPositionWithOffset = { _endSelectionPosition.X, temp2 };
+
+    // NOTE: (0,0) is top-left so vertical comparison is inverted
+    const COORD &higherCoord = (selectionAnchorWithOffset.Y <= endSelectionPositionWithOffset.Y) ? selectionAnchorWithOffset : endSelectionPositionWithOffset;
```
Comment (adiviness) [inline src/cascadia/TerminalCore/TerminalSelection.cpp]: The `&`s here are should be styled as `COORD& higherCoord`

File: src/cascadia/TerminalCore/Terminal.hpp [hunk 6]
```diff
@@ -188,5 +189,10 @@ class Microsoft::Terminal::Core::Terminal final :
 
     void _NotifyScrollEvent();
 
+    #pragma region TextSelection
+    // These methods are defined in TerminalSelection.cpp
     std::vector<SMALL_RECT> _GetSelectionRects() const;
+    const SHORT _ExpandWideGlyphSelection_Left(SHORT xPos, const SHORT yPos) const noexcept;
```
Comment (adiviness) [inline src/cascadia/TerminalCore/Terminal.hpp]: `xPos` here and below should be const.

File: src/cascadia/TerminalCore/TerminalSelection.cpp [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+// - updated x position to encapsulate the wide glyph
+const SHORT Terminal::_ExpandWideGlyphSelection_Left(const SHORT xPos, const SHORT yPos) const noexcept
+{
+    // don't change the value if at/outside the boundary
+    if (xPos <= 0 || xPos >= _buffer->GetSize().RightInclusive())
+    {
+        return xPos;
+    }
+
+    COORD position{xPos, yPos};
```
Comment (adiviness) [inline src/cascadia/TerminalCore/TerminalSelection.cpp]: nit: spacing on `{}`

File: src/cascadia/UnitTests_TerminalCore/SelectionTest.cpp [hunk 8]
```diff
@@ -178,5 +178,106 @@ namespace TerminalCoreUnitTests
                 rowValue++;
             }
         }
+
+        TEST_METHOD(SelectWideGlyph_Trailing)
+        {
+            Terminal term = Terminal();
```
Comment (DHowett-MSFT) [inline src/cascadia/UnitTests_TerminalCore/SelectionTest.cpp]: Technically this is not required; just `Terminal term;` will suffice.

File: src/cascadia/UnitTests_TerminalCore/SelectionTest.cpp [hunk 9]
```diff
@@ -178,5 +178,106 @@ namespace TerminalCoreUnitTests
...
+            DummyRenderTarget emptyRT;
+            term.Create({ 100, 100 }, 0, emptyRT);
+
+            // Insert wide glyph at position (4,10)
+            term.SetCursorPosition(4, 10);
+            term.Write(L"\xF0\x9F\x98\x83");
+
+            // Insert wide glyph at position (7,11)
+            term.SetCursorPosition(7, 11);
+            term.Write(L"\xF0\x9F\x98\x83");
```
Comment (DHowett-MSFT) [inline src/cascadia/UnitTests_TerminalCore/SelectionTest.cpp]: i am very worried about this actually: this is `00F0 009F 0098 0083`, which is four otherwise normal UTF-16 characters. You might need to formulate these as UTF-16 codepoints. How is the test passing???

File: src/cascadia/UnitTests_TerminalCore/SelectionTest.cpp [hunk 10]
```diff
@@ -178,5 +178,106 @@ namespace TerminalCoreUnitTests
...
+
+            auto viewport = term.GetViewport();
+            SHORT rowValue = 8;
+            for (auto selectionRect : selectionRects)
+            {
+                auto selection = viewport.ConvertToOrigin(selectionRect).ToInclusive();
+
+                if (rowValue == 10)
+                {
+                    VerifyCompareTraits<SMALL_RECT>::AreEqual({ 4, rowValue, 7, rowValue }, selection);
```
Comment (DHowett-MSFT) [inline src/cascadia/UnitTests_TerminalCore/SelectionTest.cpp]: you're not supposed to use `VerifyCompareTraits` directly; you're supposed to use `VERIFY_ARE_EQUAL` and friends; the traits _power_ those checkers.
